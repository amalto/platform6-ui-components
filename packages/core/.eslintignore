dist/
loader/

src/components.d.ts