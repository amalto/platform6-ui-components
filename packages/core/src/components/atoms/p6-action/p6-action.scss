$button-color: $text-strong !default;
$button-background-color: $scheme-main !default;
$button-family: false !default;
$button-border-width: $control-border-width !default;
$action-padding-vertical: calc(0.1em - #{$button-border-width}) !default;
$action-padding-horizontal: 0.2em !default;
$button-hover-color: $link-hover !default;
$button-focus-color: $link-focus !default;
$button-active-color: $link-active !default;
$button-disabled-background-color: $scheme-main !default;
$button-disabled-opacity: 0.5 !default;

@import 'src/styles/atoms/control';

:host {
  display: inline-flex;
  align-items: center;
  justify-content: center;

  button {
    @include control;
    @include unselectable;

    display: inline-flex;
    background-color: $button-background-color;
    border-width: $button-border-width;
    color: $button-color;
    cursor: pointer;
    font-family: $body-family;

    @if $button-family {
      font-family: $button-family;
    }

    justify-content: center;
    padding: $action-padding-vertical $action-padding-horizontal;
    text-align: center;
    white-space: nowrap;
    width: auto;
    height: auto;

    ::slotted(.icon:first-child:last-child),
    ::slotted(p6-icon:first-child:last-child) {
      $margin: calc(#{-1 / 2 * $action-padding-horizontal} - #{$button-border-width});

      margin-left: $margin;
      margin-right: $margin;
    }

    ::slotted(p6-icon:active),
    ::slotted(.icon:active) {
      background-attachment: grey;
      outline-color: transparent;
    }

    // States
    &:hover {
      color: $button-hover-color;
    }
    &:focus {
      color: $button-focus-color;
    }
    &:active {
      color: $button-active-color;
    }

    // Colors
    @each $name, $pair in $colors {
      $color: nth($pair, 1);
      $color-invert: nth($pair, 2);

      &.is-#{$name} {
        background-color: transparent;
        border-color: transparent; // $color;
        color: $color;

        &:hover,
        &:focus {
          background-color: transparent;
          border-color: transparent;
          color: bulmaDarken($color, 2.5%);
        }

        &:active {
          background-color: transparent;
          border-color: transparent;
          color: bulmaDarken($color, 5%);
        }

        &[disabled],
        fieldset[disabled] :host(.is-#{$name}) {
          background-color: transparent;
          border-color: transparent;
          box-shadow: none;
          color: $color;
        }

        &.is-loading {
          &::after {
            border-color: transparent transparent $color $color !important;
          }

          &:hover::after,
          &:focus::after {
            border-color: transparent transparent bulmaDarken($color, 2.5%) bulmaDarken($color, 2.5%) !important;
          }
        }
      }
    }

    // Size
    &.is-small {
      border-radius: $radius-small;
      font-size: $size-small;
    }
    &.is-normal {
      font-size: $size-normal;
    }

    &.is-loading {
      color: transparent !important;
      pointer-events: none;
      &::after {
        @include loader;
        @include center(1em);

        position: absolute !important;
      }
    }

    // Modifiers
    &[disabled],
    fieldset[disabled] :host {
      background-color: $button-disabled-background-color;
      border-color: transparent;
      opacity: $button-disabled-opacity;
    }
  }
}
