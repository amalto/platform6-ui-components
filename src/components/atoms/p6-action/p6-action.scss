@import "bulma/sass/elements/button";

@each $name, $pair in $colors {
  $color: nth($pair, 1);
  $color-invert: nth($pair, 2);
  .is-#{$name} {
    &.is-inverted {
      &.is-loading::after {
        border-color: transparent transparent $color $color !important;
      }
    }
  }
}
