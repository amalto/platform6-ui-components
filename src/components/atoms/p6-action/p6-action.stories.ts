import { Components } from "../../../components";
import {
  getElement,
  makeStory,
  ModeStoryMaker,
  Props,
  SizeStoryMaker,
} from "../../../shared/storybook/stories";
import { Mode, Size } from "../../../shared/types";

const component = "p6-action";

export default {
  title: "Atoms/Action",
  component,
};

const getStoryField = (props?: Props<Components.P6Action>): HTMLElement => {
  const size = props?.size !== undefined ? props.size : Size.normal;
  return getElement(
    component,
    getElement("p6-icon", [], {
      size,
      name: "home",
    }),
    props
  );
};

export const Default = makeStory<{
  size: Size;
  mode: Mode;
  disabled: boolean;
  waiting: boolean;
}>({
  args: {
    size: Size.normal,
    mode: Mode.default,
    disabled: false,
    waiting: false,
  },
  builder: (args) => getStoryField(args),
});

export const Modes = makeStory({
  builder: (): HTMLElement =>
    ModeStoryMaker(({ value }) =>
      getStoryField({
        mode: value,
      })
    ),
});

export const Disabled = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      disabled: true,
    }),
});

export const Waiting = makeStory({
  builder: (): HTMLElement =>
    getStoryField({
      waiting: true,
    }),
});

export const Sizes = makeStory({
  builder: (): HTMLElement =>
    SizeStoryMaker(({ value }) =>
      getStoryField({
        size: value,
      })
    ),
});
