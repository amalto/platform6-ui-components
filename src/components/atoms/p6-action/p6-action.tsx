import {
  Component,
  ComponentInterface,
  Element,
  h,
  Host,
  Prop,
} from "@stencil/core";
import { Mode, Size } from "~shared/types";
import { isEmpty } from "~utils/attribute";

@Component({
  tag: "p6-action",
  styleUrl: "p6-action.scss",
  shadow: true,
})
export class P6Action implements ComponentInterface {
  @Element() host!: HTMLP6ActionElement;

  /**
   * set the mode of the action
   */
  @Prop() mode: Mode = "default";

  /**
   * If set, shows a waiting/busy indicator
   */
  @Prop() waiting = false;

  /**
   * If `true`, the user cannot interact with the Action.
   */
  @Prop() disabled = false;

  /**
   * set the size of the action
   */
  @Prop() size: Size = "small";

  render(): JSX.Element {
    const classes = {
      "is-loading": !!this.waiting,
      [`is-${this.mode}`]: !(isEmpty(this.mode) || this.mode === "default"),
      [`is-${this.size}`]: !isEmpty(this.size) && this.size !== "default",
    };

    return (
      <Host aria-disabled={this.disabled ? "true" : null}>
        <button type="button" disabled={this.disabled} class={classes}>
          <slot />
        </button>
      </Host>
    );
  }
}
