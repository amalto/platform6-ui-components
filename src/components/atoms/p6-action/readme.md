# p6-button

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute  | Description                                          | Type                                                                     | Default     |
| ---------- | ---------- | ---------------------------------------------------- | ------------------------------------------------------------------------ | ----------- |
| `disabled` | `disabled` | If `true`, the user cannot interact with the Action. | `boolean`                                                                | `false`     |
| `mode`     | `mode`     | set the mode of the action                           | `"danger" \| "default" \| "info" \| "primary" \| "success" \| "warning"` | `"default"` |
| `size`     | `size`     | set the size of the action                           | `"default" \| "large" \| "medium" \| "normal" \| "small"`                | `"small"`   |
| `waiting`  | `waiting`  | If set, shows a waiting/busy indicator               | `boolean`                                                                | `false`     |

## Dependencies

### Used by

- [p6-translation](../../molecules/p6-translation)
- [p6-translation-line](../../molecules/p6-translation/p6-translation-line)

### Graph

```mermaid
graph TD;
  p6-translation --> p6-action
  p6-translation-line --> p6-action
  style p6-action fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
