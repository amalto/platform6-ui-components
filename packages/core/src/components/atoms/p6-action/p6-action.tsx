import { Component, ComponentInterface, Element, h, Host, JSX, Prop } from '@stencil/core';
import { Mode, Size } from '../../../shared/types';
import { getModeClass, getSizeClass } from '../../../utils/classes';

@Component({
  tag: 'p6-action',
  styleUrl: 'p6-action.scss',
  shadow: true,
})
export class P6Action implements ComponentInterface {
  @Element() host!: HTMLP6ActionElement;

  /**
   * set the mode of the action
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * If set, shows a waiting/busy indicator
   */
  @Prop() waiting = false;

  /**
   * If `true`, the user cannot interact with the Action.
   */
  @Prop() disabled = false;

  /**
   * set the size of the action
   */
  @Prop() size: Size = Size.normal;

  render(): JSX.Element {
    const classes = {
      'is-loading': !!this.waiting,
      ...getModeClass(this.mode),
      ...getSizeClass(this.size),
    };

    return (
      <Host aria-disabled={this.disabled ? 'true' : null}>
        <button type="button" disabled={this.disabled} class={classes}>
          <slot />
        </button>
      </Host>
    );
  }
}
