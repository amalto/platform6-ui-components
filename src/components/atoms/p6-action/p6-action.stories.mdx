import { library } from "@fortawesome/fontawesome-svg-core";
import { fab } from "@fortawesome/free-brands-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { action } from "@storybook/addon-actions";
import { Meta, Preview, Story } from "@storybook/addon-docs/blocks";
import { select, text, boolean, withKnobs } from "@storybook/addon-knobs";
import readme from "./readme.md";

<Meta
  title="Atoms/Action"
  component="p6-action"
  parameters={{ notes: readme }}
/>

# Action

<Story
  name="Default"
  parameters={{ docs: { disable: true } }}
  decorators={[withKnobs]}
>
  {(el) => {
    const modes = [
      "danger",
      "warning",
      "default",
      "info",
      "success",
      "primary",
    ];
    const sizes = ["small", "default", "normal", "medium", "large"];
    const kModes = select("Modes", modes, "default");
    const kSize = select("Size", sizes, "default");
    const kWaiting = boolean("Waiting", false);
    const kDisabled = boolean("Disabled", false);
    const kTransform = text("Transformation");
    const component = document.createElement("p6-action");
    component.mode = kModes;
    component.waiting = kWaiting;
    component.disabled = kDisabled;
    component.size = kSize;
    component.addEventListener("click", action("Click"));
    component.innerHTML = `<p6-icon name="home" transform=${kTransform} />`;
    return component;
  }}
</Story>

<Preview>
  <Story name="Size" height="63px">
    {`
      <p6-action size="small"><p6-icon name="home" /></p6-action>
      <p6-action><p6-icon name="home" /></p6-action>
      <p6-action size="normal"><p6-icon name="home" /></p6-action>
      <p6-action size="medium"><p6-icon name="home" /></p6-action>
      <p6-action size="large"><p6-icon name="home" /></p6-action>
    `}
  </Story>
</Preview>

<Preview>
  <Story name="Disabled" height="42px">
    {`
      <p6-action disabled><p6-icon name="home" /></p6-action>
    `}
  </Story>
</Preview>

<Preview>
  <Story name="Waiting" height="42px">
    {`
      <p6-action waiting><p6-icon name="home" /></p6-action>
    `}
  </Story>
</Preview>
