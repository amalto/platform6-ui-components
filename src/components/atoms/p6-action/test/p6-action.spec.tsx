import { newSpecPage } from "@stencil/core/testing";
import { P6Action } from "../p6-action";

describe("p6-action", () => {
  describe("mode", () => {
    const modes = ["danger", "warning", "info", "success", "primary"];

    it.each(modes)('has the class "is-%s" when mode is "%s"', async (mode) => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode=${mode}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it("does not add class when mode is default", async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode="default"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it("does not add class when mode is undefined", async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action mode=${undefined}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe("waiting", () => {
    it('has a "is-loading" class when waiting is true', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action waiting="true"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "is-loading" class when waiting is false', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action waiting="false"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe("disabled", () => {
    it('has a "disabled" attribute on the button when disabled is true', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action disabled="true"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "disabled" attribute on the button when disabled is false', async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action disabled="false"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe("size", () => {
    const sizes = ["small", "normal", "medium", "large"];

    it.each(sizes)('has the class "is-%s" when size is "%s"', async (size) => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action size=${size}></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it("does not add class when size is default", async () => {
      const page = await newSpecPage({
        components: [P6Action],
        html: `<p6-action size="default"></p6-action>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});
