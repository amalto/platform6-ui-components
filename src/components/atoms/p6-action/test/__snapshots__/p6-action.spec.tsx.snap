// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-action disabled has a "disabled" attribute on the button when disabled is true 1`] = `
<p6-action aria-disabled="true" disabled="true">
  <mock:shadow-root>
    <button class="is-small" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action disabled has no "disabled" attribute on the button when disabled is false 1`] = `
<p6-action disabled="false">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode does not add class when mode is default 1`] = `
<p6-action mode="default">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode does not add class when mode is undefined 1`] = `
<p6-action mode="undefined">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-danger" when mode is "%s" 1`] = `
<p6-action mode="danger">
  <mock:shadow-root>
    <button class="is-danger is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-info" when mode is "%s" 1`] = `
<p6-action mode="info">
  <mock:shadow-root>
    <button class="is-info is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-primary" when mode is "%s" 1`] = `
<p6-action mode="primary">
  <mock:shadow-root>
    <button class="is-primary is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-success" when mode is "%s" 1`] = `
<p6-action mode="success">
  <mock:shadow-root>
    <button class="is-small is-success" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action mode has the class "is-warning" when mode is "%s" 1`] = `
<p6-action mode="warning">
  <mock:shadow-root>
    <button class="is-small is-warning" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size does not add class when size is default 1`] = `
<p6-action size="default">
  <mock:shadow-root>
    <button type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-large" when size is "%s" 1`] = `
<p6-action size="large">
  <mock:shadow-root>
    <button class="is-large" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-medium" when size is "%s" 1`] = `
<p6-action size="medium">
  <mock:shadow-root>
    <button class="is-medium" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-normal" when size is "%s" 1`] = `
<p6-action size="normal">
  <mock:shadow-root>
    <button class="is-normal" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action size has the class "is-small" when size is "%s" 1`] = `
<p6-action size="small">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action waiting has a "is-loading" class when waiting is true 1`] = `
<p6-action waiting="true">
  <mock:shadow-root>
    <button class="is-loading is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;

exports[`p6-action waiting has no "is-loading" class when waiting is false 1`] = `
<p6-action waiting="false">
  <mock:shadow-root>
    <button class="is-small" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-action>
`;
